File: socketd/src/main/java/org/noear/socketd/protocol/CodecByteBuffer.java
package org.noear.socketd.protocol;


import org.noear.socketd.protocol.entity.EntityDefault;
import org.noear.socketd.protocol.impl.MessageDefault;

import java.nio.ByteBuffer;
import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;

/**
 * 编解码器（基于 ByteBuffer 编解）
 *
 * @author noear
 * @since 2.0
 */
public class CodecByteBuffer implements Codec<ByteBuffer> {

    private Charset charset = StandardCharsets.UTF_8;

    /**
     * 编码
     */
    @Override
    public ByteBuffer encode(Frame frame) {
        if (frame.getMessage() == null) {
            //length (flag + int.bytes)
            int len = 4 + 4;

            ByteBuffer buffer = ByteBuffer.allocate(len);

            //长度
            buffer.putInt(len);

            //flag
            buffer.putInt(frame.getFlag().getCode());

            buffer.flip();

            return buffer;
        } else {
            //key
            byte[] keyB = frame.getMessage().getKey().getBytes(charset);
            //topic
            byte[] topicB = frame.getMessage().getTopic().getBytes(charset);
            //metaString
            byte[] metaStringB = frame.getMessage().getEntity().getMetaString().getBytes(charset);

            //length (flag + key + topic + metaString + data + int.bytes + \n*3)
            int len = keyB.length + topicB.length + metaStringB.length + frame.getMessage().getEntity().getData().length + 2 * 3 + 4 + 4;

            ByteBuffer buffer = ByteBuffer.allocate(len);

            //长度
            buffer.putInt(len);

            //flag
            buffer.putInt(frame.getFlag().getCode());

            //key
            buffer.put(keyB);
            buffer.putChar('\n');

            //topic
            buffer.put(topicB);
            buffer.putChar('\n');

            //metaString
            buffer.put(metaStringB);
            buffer.putChar('\n');

            //data
            buffer.put(frame.getMessage().getEntity().getData());

            buffer.flip();

            return buffer;
        }
    }

    /**
     * 解码
     */
    @Override
    public Frame decode(ByteBuffer buffer) {
        int len0 = buffer.getInt();

        if (len0 > (buffer.remaining() + 4)) {
            return null;
        }

        int flag = buffer.getInt();

        if (len0 == 8) {
            //len + flag
            return new Frame(Flag.Of(flag), null);
        } else {

            //1.解码key and topic
            ByteBuffer sb = ByteBuffer.allocate(Math.min(4096, buffer.limit()));

            //key
            String key = decodeString(buffer, sb, 256);
            if (key == null) {
                return null;
            }

            //topic
            String topic = decodeString(buffer, sb, 512);
            if (topic == null) {
                return null;
            }

            //metaString
            String metaString = decodeString(buffer, sb, 4096);
            if (metaString == null) {
                return null;
            }

            //2.解码 body
            int len = len0 - buffer.position();
            byte[] data = new byte[len];
            if (len > 0) {
                buffer.get(data, 0, len);
            }

            MessageDefault message = new MessageDefault().key(key).topic(topic).entity(new EntityDefault(metaString, data));
            message.flag(Flag.Of(flag));
            return new Frame(message.getFlag(), message);
        }
    }

    protected String decodeString(ByteBuffer buffer, ByteBuffer sb, int maxLen) {
        sb.clear();

        while (true) {
            byte c = buffer.get();

            if (c == 10) { //10:'\n'
                break;
            } else if (c != 0) { //32:' '
                sb.put(c);
            }

            if (maxLen > 0 && maxLen < sb.position()) {
                return null;
            }
        }

        sb.flip();
        if (sb.limit() < 1) {
            return "";
        }

        return new String(sb.array(), 0, sb.limit());
    }
}

File: python/socketd/socketd/transport/core/codec/CodecByteBuffer.py
from io import BytesIO
from typing import Callable, Optional
from socketd.transport.core.Costants import Constants, Flag
from socketd.transport.core.Frame import Frame
from socketd.transport.core.entity.MessageDefault import MessageDefault
from socketd.transport.core.entity.EntityDefault import EntityDefault
from socketd.transport.core.Config import Config
from socketd.transport.core.codec.Buffer import Buffer
from socketd.transport.core.Codec import Codec, CodecReader, CodecWriter


def assert_size(name: str, size: int, limitSize: int) -> None:
    if size > limitSize:
        buf = f"This message {name} size is out of limit {limitSize} ({size})"
        raise RuntimeError(buf)


class ByteBufferCodecReader(CodecReader):

    def __init__(self, buffer: Buffer):
        self.__buffer = buffer

    def get_bytes(self, size) -> bytes:
        return self.__buffer.read(size)

    def get_int(self) -> int:
        return int.from_bytes(self.__buffer.read1(4), byteorder='big', signed=False)

    def skip_bytes(self, size):
        self.__buffer.seek(self.__buffer.tell() + size)

    def seek(self, size):
        self.__buffer.seek(size)

    def remaining(self):
        return self.__buffer.remaining()

    def position(self):
        return self.__buffer.tell()

    def get_buffer(self) -> BytesIO:
        return self.__buffer

    def close(self):
        self.__buffer.close()


class ByteBufferCodecWriter(CodecWriter):

    def __init__(self, buffer: Buffer):
        self.__buffer = buffer

    def put_bytes(self, _bytes: bytearray | memoryview | bytes):
        self.__buffer.write(_bytes)

    def put_int(self, _num: int):
        self.__buffer.write(_num.to_bytes(length=4, byteorder='big', signed=False))

    def put_char(self, _val: bytes):
        by: int = int.from_bytes(_val, byteorder='big', signed=False)
        self.__buffer.write(by.to_bytes(length=2,byteorder='big', signed=False))
        # self.__buffer.write(_val)

    def flush(self):
        self.__buffer.flush()

    def get_buffer(self) -> Buffer:
        return self.__buffer

    def close(self):
        self.__buffer.close()


class CodecByteBuffer(Codec):
    def __init__(self, config: Config):
        self.config = config

    def write(self, frame: Frame, factory: Callable[[int], CodecWriter]) -> CodecWriter:
        if frame.message is None:
            # length (flag + int.bytes)
            _len = 2 * 4
            target: CodecWriter = factory(_len)
            # length
            target.put_int(_len)
            # flag
            target.put_int(frame.flag)
            target.flush()

            return target
        else:
            # sid
            sidB: bytes = frame.message.get_sid().encode(self.config.get_charset())
            # event
            event: bytes = frame.message.get_event().encode(self.config.get_charset())
            # metaString
            metaStringB: bytes = frame.message.get_entity().get_meta_string().encode(self.config.get_charset())

            # length (flag + sid + event + metaString + data + int.bytes + \n*3)
            len1 = len(sidB) + len(event) + len(
                metaStringB) + frame.message.get_entity().get_data_size() + 1 * 3 + 2 * 4

            assert_size("sid", len(sidB), Constants.MAX_SIZE_SID)
            assert_size("event", len(event), Constants.MAX_SIZE_EVENT)
            assert_size("metaString", len(metaStringB), Constants.MAX_SIZE_META_STRING)
            assert_size("data", frame.message.get_entity().get_data_size(), Constants.MAX_SIZE_FRAGMENT)

            target: CodecWriter = factory(len1)

            # length
            target.put_int(len1)

            # flag
            target.put_int(frame.flag)

            # sid
            target.put_bytes(sidB)
            target.put_char(b'\n')

            # event
            target.put_bytes(event)
            target.put_char(b'\n')

            # metaString
            target.put_bytes(metaStringB)
            target.put_char(b'\n')

            # _data
            if frame.message.get_entity().get_data() is not None:
                target.put_bytes(frame.message.get_entity().get_data().getvalue())
            target.flush()

            return target

    def read(self, _reader: CodecReader) -> Frame | None:
        len0 = _reader.get_int()

        if len0 > (_reader.remaining() + 4):
            return None
        flag = _reader.get_int()  # 取前一位数据

        if len0 == 8:
            # len + flag
            return Frame(Flag.of(flag), None)
        else:
            metaBufSize = min(Constants.MAX_SIZE_META_STRING, _reader.remaining())
            # 1. decode sid and event
            by = Buffer(limit=metaBufSize)
            sid = self.decodeString(_reader, by, Constants.MAX_SIZE_SID)
            event = self.decodeString(_reader, by, Constants.MAX_SIZE_EVENT)
            metaString = self.decodeString(_reader, by, Constants.MAX_SIZE_META_STRING)

            # 2. decode body
            dataRealSize = len0 - _reader.position() + 3
            data: Optional[bytearray] = None
            if dataRealSize > Constants.MAX_SIZE_FRAGMENT:
                # exceeded the limit, read and discard the bytes
                data = bytearray(Constants.MAX_SIZE_FRAGMENT)
                _reader.get_buffer().readinto(data)
                for i in range(dataRealSize - Constants.MAX_SIZE_FRAGMENT):
                    _reader.get_bytes(1)
            else:
                data = bytearray(_reader.get_bytes(dataRealSize))

            message = MessageDefault().set_sid(sid).set_event(event).set_entity(
                EntityDefault().meta_string_set(metaString).data_set(data)
            )
            message.flag = Flag.of(flag)
            _reader.close()
            return Frame(message.flag, message)

    def decodeString(self, reader: CodecReader, buf: Buffer, maxLen: int) -> str:
        b = bytearray(reader.get_buffer().readline(maxLen))[:-2]
        if buf.limit() < 1:
            return ""
        return b.decode(self.config.get_charset())
